Add todo app with task model and integrate todo app APIs with projects part(including PAF process, contracting and invoicing)

sandeepsajan0 opened the pull request:
Fixes #3628 Closes #3605 

After:
![image](https://github.com/HyphaApp/hypha/assets/23638629/01349224-885d-4fd9-9d92-c3e8719f9d40)

sandeepsajan0 committed:
Initialize task app, initial models and methods

theskumar commented on hypha/apply/tasks/models.py:
```diff
@@ -0,0 +1,90 @@
+        related_name="task_related",
+    )
+    related_object_id = models.PositiveIntegerField(blank=True, null=True)
+    related_object = GenericForeignKey("related_content_type", "related_object_id")
```
Thoughts on linking an `activity` object that lead to this task. Can't think of an use-case now, but it will be use to track the source activity that lead to these actionable items.

theskumar commented on hypha/settings/django.py:
```diff
@@ -18,6 +18,7 @@
     "hypha.apply.review",
     "hypha.apply.determinations",
     "hypha.apply.stream_forms",
+    "hypha.apply.tasks",
```
Please rename it to `user_tasks` or simply `todo` app. `tasks` too generic

theskumar replied:
Or `user_actions`

sandeepsajan0 replied:
Renamed it to `todo`. `todo` makes more sense than others.

sandeepsajan0 committed:
WIP: Fixing validations, add todo list

sandeepsajan0 committed:
Add Validation and List tasks for users

Update validations to save task \n Add icons for template map and sprites \n Update text link style to markdown \n List tasks for staff(if exists) \n Fix template_map dict update issue(deepcopy)

sandeepsajan0 committed:
Fix APIs for user groups, and fix union issue

sandeepsajan0 committed:
APIs are tested and ready to integrate.

sandeepsajan0 committed:
Integrate task apis to project part

sandeepsajan0 committed:
Renamed tasks app to todo app

sandeepsajan0 committed:
Handle invoice delete case

sandeepsajan0 committed:
Refactor task update code for invocies, use services

sandeepsajan0 committed:
Fixed Group import

frjo commented:
@sandeepsajan0 Was this ready for testing?

sandeepsajan0 commented:
@frjo Yes, it is ready for testing.

theskumar commented on hypha/apply/todo/views.py:
```diff
@@ -0,0 +1,138 @@
+        related_obj: Object - Submission, Project, Invoice, Report
+    output: None
+    """
+    task = Task.objects.filter(
+        code=code,
+        user=user,
+        related_content_type=ContentType.objects.get_for_model(related_obj).id,
+        related_object_id=related_obj.id,
+    ).first()
+    if task:
+        task.delete()
```
```suggestion
    Task.objects.filter(
        code=code,
        user=user,
        related_content_type=ContentType.objects.get_for_model(related_obj).id,
        related_object_id=related_obj.id,
    ).delete()
```

Consider using a more optimized version of delete

theskumar commented on hypha/apply/todo/views.py:
```diff
@@ -0,0 +1,138 @@
+    user_group_matching_tasks = matching_tasks.annotate(
+        group_count=Count("user_group")
+    ).filter(group_count=len(user_group.all()))
+    for group in user_group.all():
+        user_group_matching_tasks = user_group_matching_tasks.filter(
+            user_group__id=group.id
+        )
+    if user_group_matching_tasks.exists():
+        user_group_matching_tasks.delete()
```
This looks is over-writing the `user_group_matching_tasks` in every iteration. You are then removing the last one. Something is not adding up. 

Would be great if you can write a test to check it's going what it should do.

sandeepsajan0 replied:
It is just finding the tasks with exact user groups(many to many relations) and in the end, deleting those tasks.

I'll add tests for all of these APIs later.  As this PR already has too many changes, I'll add a separate PR for that.

theskumar replied:
It's okay to have the PR as part of this PR itself. Also can you clarify bit more what you mean by `in the end`. 

You are looping over something and just picking the last value for that iteration. I don't see it being predictable. 

image user_group.all() returns `group 1` and `group 2`, you are going be acting on the `group 2` and not do anything for `group 1`. If that is case, you can just `user_group.last()` instead of `user_group.all()`

sandeepsajan0 replied:
Okay, Let me explain the loop logic here. As I mentioned in my last comment it aims to find the tasks with 'exact' user_groups.

Ex: Let's say there are two groups `group1` and `group2` in user_group.
**In line 82:** annotate and filter is making sure that `user_group_matching_tasks` contains all the tasks with two groups(len of 'user_group'), but we are not sure these groups are exclusively `group1` and `group2`.
So for that, we iterate through 'user_group'. In the first iteration, the code filters `user_group_matching_tasks` to include only those with `group1` and updates the collection accordingly. In the second iteration, it ensures that the 'updated' `user_group_matching_tasks` collection also contains `group2`. 

After all iterations, we can be sure that all tasks in `user_group_matching_tasks` have exactly two groups, specifically `group1` and `group2`.

theskumar commented on hypha/apply/todo/views.py:
```diff
@@ -0,0 +1,138 @@
+    user_group_tasks = Task.objects.annotate(group_count=Count("user_group")).filter(
+        group_count=len(user.groups.all())
+    )
+    for group in user.groups.all():
+        user_group_tasks = user_group_tasks.filter(user_group__id=group.id)
+
```
Same comment here as https://github.com/HyphaApp/hypha/pull/3634/files#r1421747171

wes-otf commented:
In testing this worked really well! Only question I had was if we could potentially make the tasks stand out like then do in your screenshot @sandeepsajan0 by changing the background or task color (ie. what was done in #3651) or give them defined borders. 

With contrast (prev screenshot):

![285852704-01349224-885d-4fd9-9d92-c3e8719f9d40](https://github.com/HyphaApp/hypha/assets/145372368/c4d0bee3-5095-4e0c-b6eb-96d0b92c90f2)

In test:

<img width="1183" alt="Screenshot 2023-12-15 at 11 56 11" src="https://github.com/HyphaApp/hypha/assets/145372368/26deecc3-786f-4982-a501-6974a69a82bf">

Otherwise this is ready for live!

sandeepsajan0 commented:
@wes-otf I get your point. We need to update the background color for all users' dashboards, we have already updated it for applicants in #3602. I'll add a PR for other dashboards as well.
